### PR TITLE
Optimize `update_snapshot()`

### PR DESCRIPTION
**Pull request by RasmusSkytte**
### Intent
This PR introduces optimisations to speed up `update_snapshot()`

(Performance with low-level SQL generation)
billede

From benchmarking, it seems that performance was gained for SQLite by adding the indexes.
For the other backends, the performance was gained from using the lower level SQL generation.

### Approach
* For tables created by `create_table()`, we now also add a unique index for (checksum, from_ts) since they are unique during `update_snapshot()` operations.

* The operations in `update_snapshot()` was changed to use SQL directly in many cases.
  Often, the actual SQL is still generated by `dbplyr`.

* A message is added when connecting with `odbc` driver informing about the `odbc.batch_rows` option.

* The tests for `update_snapshot()` only partially tested the logs being written with a configured logger.
  We now perform additional tests on the output to the database log
   

### Known issues
N/A

### Checklist

* [x] The PR passes all local unit tests
* [x] I have documented any new features introduced
* [x] If the PR adds a new feature, please add an entry in `NEWS.md`
* [x] A reviewer is assigned to this PR

### CODE REVIEW

**Comment by RasmusSkytte** [comment]
Notes for later:

`$log_to_db()` calls should now match the previous version of `update_snapshot()`. 

This version of `update_snapshot()` is still missing the `$log_info()` calls of the previous version. We should discuss whether to keep the same structure of this logging or make changes now.

**Comment by RasmusSkytte** [comment]
Notes for later:

After adding the `$log_to_db()` calls, our performance has dropped in the benchmarks:

billede

**Comment by RasmusSkytte** [comment]
Notes for later:

Using more low level SQL generation means we can save some computes:

billede

File: tests/testthat/test-update_snapshot.R [hunk 1]
```diff
@@ -85,37 +76,177 @@ test_that("update_snapshot() works", {
...
 
+    # Check the content of the log table
+    expect_identical(as.character(logs$date), as.character(timestamp))
+
+    db_table_id <- id(db_table, conn)
+    if ("catalog" %in% colnames(logs)) expect_identical(logs$catalog, purrr::pluck(db_table_id, "name", "catalog"))
+    expect_identical(logs$schema, purrr::pluck(db_table_id, "name", "schema"))
+    expect_identical(logs$table, purrr::pluck(db_table_id, "name", "table"))
+
+    expect_identical(logs$n_insertions, 32L)
```
Comment (LasseEngboChr) [inline tests/testthat/test-update_snapshot.R:93]: Here we condition on mtcars never changing ... is it safe or should we use nrow(.data)?

File: tests/testthat/test-update_snapshot.R [hunk 2]
```diff
@@ -85,37 +76,177 @@ test_that("update_snapshot() works", {
...
+      nrow(slice_time(target, "2022-10-01 09:00:00")),
+      nrow(mtcars)
+    )
+
+    expect_identical(
+      slice_time(target, "2022-10-03 09:00:00") |>
+        dplyr::select(!c("from_ts", "until_ts", "checksum")) |>
+        dplyr::collect() |>
+        dplyr::arrange(wt, qsec),
+      .data |>
```
Comment (LasseEngboChr) [inline tests/testthat/test-update_snapshot.R:154]: A little confusing that you change .data instead of creating .data_modified

Comment (RasmusSkytte) [reply tests/testthat/test-update_snapshot.R:154]: I am not changing the data only the order. The order of the records is not (always) respected when using databases, so to compare that the data are equal, I need to sort them first to create the same order for both.

**Review by LasseEngboChr — APPROVED** [review]

File: tests/testthat/test-update_snapshot.R [hunk 3]
```diff
@@ -401,6 +401,53 @@ test_that("update_snapshot() works (holistic test 2)", {
...
+      log_conn = conn,
+      output_to_console = FALSE
+    )
+
+    # Update the table with update_snapshot() and store the results
+    update_snapshot(.data, conn, db_table, "2022-10-03 09:00:00", logger = create_logger("2022-10-03 09:00:00"))
+    target_data_1 <- get_table(conn, db_table) |> dplyr::collect()
+
+    # Update the table with the same data again update_snapshot() and store the results
+    update_snapshot(.data, conn, db_table, "2022-10-04 09:00:00", logger = create_logger("2022-10-03 09:00:00"))
```
Comment (LasseEngboChr) [inline tests/testthat/test-update_snapshot.R:432]: Should be same time

**Review by LasseEngboChr — APPROVED** [review]

**Pull request merged**